scripts/asset_processor/ffmpeg_capture.py

```diff
@@ -0,0 +1,241 @@
+		if 'opencv' in video_reader:
+			global cv2
+			import cv2
+		self._read_metadata()
+
+	def _read_metadata(self):
+		"""
+		Reads video properties and fills corresponding fields
+		@return:
+		"""
+		cap = None
+		try:
+			cap = cv2.VideoCapture(self.filename, cv2.CAP_FFMPEG)
+			self.fps = cap.get(cv2.CAP_PROP_FPS)
+			self.width = int(cap.get(cv2.CAP_PROP_FRAME_WIDTH))
+			self.height = int(cap.get(cv2.CAP_PROP_FRAME_HEIGHT))
+			self.frame_count = int(cap.get(cv2.CAP_PROP_FRAME_COUNT))
+			if self.video_reader == 'opencv' or self.video_reader == 'opencv_pts':
+				self.opencv_capture = cap
+			logger.info(f'Video file opened {self.filename}, {self.width}x{self.height}, {self.fps} FPS')
+		finally:
+			if cap is not None and self.video_reader != 'opencv' and self.video_reader != 'opencv_pts':
+				cap.release()
+
+	def _read_next_frame(self):
+		if self.video_reader == 'ffmpeg_yuv':
+			# get raw frame from stdout and convert it to numpy array
+			bytes = self.video_capture.stdout.read(int(self.height * self.width * 6 // 4))
+			if len(bytes) == 0:
+				return None
+			k = self.height * self.width
+			y = np.frombuffer(bytes[0:k], dtype=np.uint8).reshape((self.height, self.width))
+			u = np.frombuffer(bytes[k:k + k // 4], dtype=np.uint8).reshape((self.height // 2, self.width // 2))
+			v = np.frombuffer(bytes[k + k // 4:], dtype=np.uint8).reshape((self.height // 2, self.width // 2))
+			u = np.reshape(cv2.resize(np.expand_dims(u, -1), (self.width, self.height)), (self.height, self.width))
+			v = np.reshape(cv2.resize(np.expand_dims(v, -1), (self.width, self.height)), (self.height, self.width))
+			return self.pixel_converter.convert([y], [u], [v])[0]
+		elif self.video_reader == 'ffmpeg_bgr':
+			bytes = self.video_capture.stdout.read(int(self.height * self.width * 3))
+			if len(bytes) == 0:
+				return None
+			return np.frombuffer(bytes, np.uint8).reshape([self.height, self.width, 3])
+		elif self.video_reader == 'opencv' or self.video_reader == 'opencv_pts':
+			return self.opencv_capture.read()[1]
+
+	def read(self) -> Union[FrameData, None]:
+		"""
+		Reads next frame from video.
+		@return: Tuple[frame_index, frame_timestamp, frame] or [None, None, None] if end of video
+		"""
+		if not self.started:
+			self.start()
+		frame = self._read_next_frame()
+		if frame is None:
+			return None
+		self.frame_idx += 1
+		if 0 < self.frame_count == self.frame_idx:
+			logger.error(f'Frame count mismatch, possibly corrupted video file: {self.filename}')
+			self.release()
+			return None
+		timestamp = self._get_timestamp_for_frame(self.frame_idx)
+		logger.debug(f'Read frame {self.frame_idx} at PTS_TIME {timestamp}')
+		return FfmpegCapture.FrameData(self.frame_idx, timestamp, frame)
+
+	def _get_timestamp_for_frame(self, frame_idx) -> float:
+		if self.video_reader == 'opencv_pts':
+			opencv_ts = self.opencv_capture.get(cv2.CAP_PROP_POS_AVI_RATIO) * self.opencv_capture.get(cv2.CAP_PROP_POS_MSEC)
```

Comment:
Why do we have to multiply `CAP_PROP_POS_MSEC` by `CAP_PROP_POS_AVI_RATIO`? Wouldn't `CAP_PROP_POS_MSEC` be the PTS in milliseconds?

Reply:
Originally, it's just time calculated from N frame and FPS. In a custom build, it returns raw PTS from Ffmpeg, which is in CAP_PROP_POS_AVI_RATIO units. There are special constants related to this property, which may be useful at some point.

Reply:
Got it. Thanks for the clarification.